src/java.instrument/share/classes/sun/instrument/InstrumentationImpl.java

```diff
@@ -470,45 +454,18 @@ public Object run() {
             try {
                 m = javaAgentClass.getDeclaredMethod(methodname,
                                                  new Class<?>[] { String.class });
-            } catch (NoSuchMethodException x) {
-                // ignore this exception because we'll try
-                // two arg inheritance next
-            }
-        }
-
-        if (m == null) {
-            // now try the inherited 2-arg method
-            try {
-                m = javaAgentClass.getMethod( methodname,
-                                 new Class<?>[] {
-                                     String.class,
-                                     java.lang.instrument.Instrumentation.class
-                                 }
-                               );
-                twoArgAgent = true;
-            } catch (NoSuchMethodException x) {
-                // ignore this exception because we'll try
-                // one arg inheritance next
-            }
-        }
-
-        if (m == null) {
-            // finally try the inherited 1-arg method
-            try {
-                m = javaAgentClass.getMethod(methodname,
-                                             new Class<?>[] { String.class });
             } catch (NoSuchMethodException x) {
                 // none of the methods exists so we throw the
                 // first NoSuchMethodException as per 5.0
                 throw firstExc;
             }
         }
 
-        // the premain method should not be required to be public,
-        // make it accessible so we can call it
-        // Note: The spec says the following:
-        //     The agent class must implement a public static premain method...
-        setAccessible(m, true);
+        // reject non-public premain or agentmain method
+        if (!m.canAccess(null)) {
+            String msg = "method " + classname + "." +  methodname + " must be declared public";
+            throw new IllegalAccessException(msg);
+        }
```

Comment:
I think the updated patch looks much better. 

If the agent class doesn't declare a premain then NoSuchMethodException will be thrown.

The only thing that I'm wondering about now is the exception message when the agent class is not public. If I read the changes correctly it means that IllegalAccessException will be thrown saying that <class>.premain should be public. Is that correct? We might need to adjust to the exception message to make it clear, or put in an explicit then to ensure to test that the agent class is public.

Reply:
I think an explicit check for a public premain method is better to disambiguate the cases and provide appropriate error messages.